File: frontend/modules/js/busPasses.js
import { Modal } from "/static/common/modal.js";
import { request } from "/static/common/api.js";
import { createDropdown } from "/static/common/dropdown.js";
import { confirmDeletion, showOperationFeedback } from "/static/common/operationFeedback.js";

const API = "/bus-passes";

function escapeHtml(value) {
    return String(value ?? "").replace(/[&<>'"]/g, character => ({
        "&": "&amp;", "<": "&lt;", ">": "&gt;", "'": "&#39;", '"': "&quot;"
    }[character]));
}

function dateValue(value) {
    return value ? String(value).slice(0, 10) : "";
}

function formatDate(value) {
    if (!value) return "—";
    const date = new Date(`${dateValue(value)}T00:00:00`);
    return Number.isNaN(date.getTime()) ? escapeHtml(value) : new Intl.DateTimeFormat(undefined, {
        day: "2-digit", month: "short", year: "numeric"
    }).format(date);
}

function calendarDate(value) {
    return new Date(`${value}T00:00:00`);
}

function dateInputValue(date) {
    const year = date.getFullYear();
    const month = String(date.getMonth() + 1).padStart(2, "0");
    const day = String(date.getDate()).padStart(2, "0");
    return `${year}-${month}-${day}`;
}

function plannedExpiry(startValue, period) {
    const start = calendarDate(startValue);
    if (Number.isNaN(start.getTime()) || period === "Custom Dates") return "";
    if (period === "One Day") return dateInputValue(start);
    const originalMonth = start.getMonth();
    start.setFullYear(start.getFullYear() + 1);
    // Match the server's leap-day rule: 29 Feb renews on 28 Feb.
    if (start.getMonth() !== originalMonth) start.setDate(0);
    return dateInputValue(start);
}

function badge(pass) {
    if (!pass) return '<span class="bus-passes-badge is-pending">Not issued</span>';
    const status = pass.effective_status || pass.status || "Pending";
    return `<span class="bus-passes-badge is-${status.toLowerCase()}">${escapeHtml(status)}</span>`;
}

function recordRow(record) {
    const pass = record.bus_pass;
    const assignment = record.transport;
    const action = pass
        ? `<div class="bus-passes-actions"><button type="button" class="bus-passes-action edit-pass" data-pass-id="${pass.id}">Manage</button><button type="button" class="bus-passes-action is-delete delete-pass" data-pass-id="${pass.id}">Delete</button></div>`
        : `<button type="button" class="bus-passes-action issue-pass" data-student-id="${record.student.id}" ${assignment.assigned ? "" : "disabled title=\"Assign a route and bus first\""}>Issue pass</button>`;
    return `<tr>
        <td><strong>${escapeHtml(record.student.name)}</strong><small>${escapeHtml(record.student.student_code)}</small></td>
        <td>${assignment.assigned ? `<strong>${escapeHtml(assignment.bus_number)}</strong><small>${escapeHtml(assignment.route_name)}</small>` : '<span class="bus-passes-unassigned">Bus assignment pending</span>'}</td>
        <td>${escapeHtml(assignment.boarding_stop || "—")}</td>
        <td>${pass ? `<strong>${escapeHtml(pass.pass_number)}</strong><small>${escapeHtml(pass.validity_period)}</small>` : "—"}</td>
        <td>${badge(pass)}${pass?.expiring_soon ? '<small class="bus-passes-expiry">Expires within 30 days</small>' : ""}</td>
        <td>${pass ? formatDate(pass.valid_until) : "—"}</td>
        <td>${action}</td>
    </tr>`;
}

function renderPage(data) {
    const root = document.createElement("section");
    root.className = "bus-passes-page";
    root.innerHTML = `
        <header class="bus-passes-hero glass-card">
            <div><p class="bus-passes-eyebrow">Transport administration</p><h1>Bus Passes</h1><p>Issue, renew, and suspend passes. The bus, route, and boarding stop always come from each student's central transport assignment.</p></div>
            <div class="bus-passes-summary"><strong>${data.issued}/${data.total}</strong><span>passes issued</span><small>${data.expiring_soon} expiring within 30 days</small></div>
        </header>
        <section class="bus-passes-guidance glass-card"><span>1. Assign bus and route</span><i>→</i><span>2. Assign student and stop</span><i>→</i><span>3. Issue pass for any validity</span><i>→</i><span>4. Renew or change dates</span></section>
        <section class="bus-passes-table-card glass-card">
            <div class="bus-passes-table-heading"><div><h2>Student bus passes</h2><p>Set one-day, yearly, two-semester, or custom expiry dates directly when issuing or renewing.</p></div><button type="button" class="bus-passes-refresh">↻ Refresh</button></div>
            <div class="bus-passes-table-wrap"><table class="bus-passes-table"><thead><tr><th>Student</th><th>Assigned transport</th><th>Boarding stop</th><th>Pass number</th><th>Status</th><th>Valid until</th><th>Action</th></tr></thead><tbody>${data.records.length ? data.records.map(recordRow).join("") : '<tr><td colspan="7" class="bus-passes-empty">No student accounts are available.</td></tr>'}</tbody></table></div>
        </section>`;
    attachEvents(root, data);
    return root;
}

function formMarkup(record, mode) {
    const pass = record.bus_pass;
    const today = new Date().toISOString().slice(0, 10);
    const initialStartDate = dateValue(pass?.valid_from) || today;
    const initialValidityPeriod = pass?.validity_period || "One Year";
    // A normal pass is always shown as one full year by default.  Only a
    // deliberately custom pass keeps its saved end date unchanged.
    const initialExpiryDate = initialValidityPeriod === "Custom Dates"
        ? (dateValue(pass?.valid_until) || initialStartDate)
        : (plannedExpiry(initialStartDate, initialValidityPeriod) || initialStartDate);
    return `<form class="bus-pass-form">
        <p class="bus-pass-form__assignment"><strong>${escapeHtml(record.student.name)}</strong><br>${escapeHtml(record.transport.bus_number || "Bus assignment pending")} · ${escapeHtml(record.transport.route_name || "No route")}</p>
        <div class="modal-group"><label class="modal-label" for="bus-pass-valid-from">${mode === "issue" ? "Pass start date" : "Renewal start date"}</label><input id="bus-pass-valid-from" type="date" value="${initialStartDate}" required></div>
        <div class="modal-group"><label class="modal-label">Validity plan</label><div class="bus-pass-period-dropdown"></div><small class="bus-pass-form__hint">One Day ends on the selected day. You may always edit the expiry date.</small></div>
        <div class="modal-group"><label class="modal-label" for="bus-pass-valid-until">Expiry date</label><input id="bus-pass-valid-until" type="date" value="${initialExpiryDate}" required></div>
        <div class="modal-group"><label class="modal-label" for="bus-pass-academic-year">Academic year / session</label><input id="bus-pass-academic-year" type="text" maxlength="30" value="${escapeHtml(pass?.academic_year || "")}" placeholder="2026–2027"></div>
        <div class="modal-group"><label class="modal-label">Pass status</label><div class="bus-pass-status-dropdown"></div></div>
    </form>`;
}

function openPassModal(record, mode, reload) {
    const form = document.createElement("div");
    form.innerHTML = formMarkup(record, mode);
    const pass = record.bus_pass;
    const periodDropdown = createDropdown({
        id: "bus-pass-period",
        placeholder: "Select validity",
        value: pass?.validity_period || "One Year",
        items: ["One Day", "One Year", "Two Semesters", "Custom Dates"],
    });
    const statusDropdown = createDropdown({
        id: "bus-pass-status",
        placeholder: "Select status",
        value: pass?.status || "Active",
        items: mode === "issue" ? ["Active", "Pending"] : ["Active", "Pending", "Suspended"],
    });
    form.querySelector(".bus-pass-period-dropdown").appendChild(periodDropdown);
    form.querySelector(".bus-pass-status-dropdown").appendChild(statusDropdown);

    const validFrom = form.querySelector("#bus-pass-valid-from");
    const validUntil = form.querySelector("#bus-pass-valid-until");
    const applyPlannedExpiry = () => {
        const expiry = plannedExpiry(validFrom.value, periodDropdown.getValue());
        if (expiry) validUntil.value = expiry;
    };
    periodDropdown.onChange(applyPlannedExpiry);
    validFrom.addEventListener("change", applyPlannedExpiry);
    validUntil.addEventListener("change", () => {
        const planned = plannedExpiry(validFrom.value, periodDropdown.getValue());
        if (planned && validUntil.value !== planned) periodDropdown.setValue("Custom Dates");
    });
    const submitText = mode === "issue" ? "Issue bus pass" : "Save pass changes";
    Modal.form({
        eyebrow: "BUS PASS MANAGEMENT",
        title: mode === "issue" ? `Issue pass for ${record.student.name}` : `Manage ${record.student.name}'s pass`,
        subtitle: mode === "issue" ? "The server creates the pass number; you control the validity dates." : "Renew, suspend, or change any pass date as needed.",
        content: form,
        submitText,
        onSubmit: async () => {
            const payload = {
                valid_from: validFrom.value,
                valid_until: validUntil.value,
                validity_period: periodDropdown.getValue(),
                academic_year: form.querySelector("#bus-pass-academic-year").value.trim() || null,
                status: statusDropdown.getValue(),
            };
            try {
                if (mode === "issue") {
                    await request(API, { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify({ student_id: record.student.id, ...payload }) });
                } else {
                    await request(`${API}/${record.bus_pass.id}`, { method: "PUT", headers: { "Content-Type": "application/json" }, body: JSON.stringify(payload) });
                }
                Modal.close();
                await reload();
                showOperationFeedback({
                    type: "success",
                    title: mode === "issue" ? "Bus pass issued" : "Bus pass updated",
                    message: mode === "issue" ? "The student can now view the issued pass." : "The pass changes have been saved."
                });
            } catch (error) {
                showOperationFeedback({
                    type: "error",
                    title: "Bus pass not saved",
                    message: error.message || "Please review the pass details and try again."
                });
                throw error;
            }
        },
    });
}

function attachEvents(root, data) {
    const reload = async () => {
        root.replaceWith(await loadPage());
    };
    root.querySelector(".bus-passes-refresh").addEventListener("click", () => void reload());
    root.querySelectorAll(".issue-pass").forEach(button => button.addEventListener("click", () => {
        const record = data.records.find(item => item.student.id === Number(button.dataset.studentId));
        if (record) openPassModal(record, "issue", reload);
    }));
    root.querySelectorAll(".edit-pass").forEach(button => button.addEventListener("click", () => {
        const record = data.records.find(item => item.bus_pass?.id === Number(button.dataset.passId));
        if (record) openPassModal(record, "manage", reload);
    }));
    root.querySelectorAll(".delete-pass").forEach(button => button.addEventListener("click", async () => {
        const record = data.records.find(item => item.bus_pass?.id === Number(button.dataset.passId));
        if (!record) return;
        const confirmed = await confirmDeletion({
            title: "Delete this bus pass?",
            message: `This removes ${record.student.name}'s pass but keeps their bus, route, and stop assignment unchanged.`,
            confirmText: "Delete pass"
        });
        if (!confirmed) return;
        try {
            await request(`${API}/${record.bus_pass.id}`, { method: "DELETE" });
            await reload();
            showOperationFeedback({
                type: "success",
                title: "Bus pass deleted",
                message: "The student's transport assignment was kept unchanged."
            });
        } catch (error) {
            showOperationFeedback({
                type: "error",
                title: "Bus pass not deleted",
                message: error.message || "Please try again."
            });
        }
    }));
}

async function loadPage() {
    const data = await request(API);
    return renderPage(data);
}

export function render() {
    const loading = document.createElement("section");
    loading.className = "bus-passes-loading";
    loading.textContent = "Loading bus passes…";
    void loadPage().then(page => loading.replaceWith(page)).catch(error => {
        loading.innerHTML = `<section class="bus-passes-error"><h2>Bus passes unavailable</h2><p>${escapeHtml(error.message)}</p></section>`;
    });
    return loading;
}

File: backend/routes/bus_passes.py
"""Admin management for student transport passes.

Passes never own a bus, route, or stop.  Those values are always resolved from
the existing Student assignment so Admin, Student, and Live Tracking screens
cannot drift apart.
"""

from datetime import date, datetime, timezone
import secrets

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from backend.database import get_db
from backend.models import Bus, BusPass, Route, Student, User
from backend.schemas import BusPassIssue, BusPassUpdate
from backend.security import require_management


router = APIRouter(prefix="/api/bus-passes", tags=["Bus Passes"])


def _one_year_after(start: date) -> date:
    """Return the same calendar day in the next year, safely handling Feb 29."""

    try:
        return start.replace(year=start.year + 1)
    except ValueError:
        return start.replace(year=start.year + 1, month=2, day=28)


def _resolve_valid_until(
    valid_from: date,
    validity_period: str,
    requested_valid_until: date | None,
) -> date:
    """Use the requested expiry when supplied, otherwise apply the selected plan."""

    if requested_valid_until is not None:
        valid_until = requested_valid_until
    elif validity_period == "One Day":
        # A one-day pass starts and ends on the purchased calendar day.
        valid_until = valid_from
    elif validity_period in {"One Year", "Two Semesters"}:
        valid_until = _one_year_after(valid_from)
    else:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Choose an expiry date for a custom-validity pass.",
        )

    if valid_until < valid_from:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="The expiry date cannot be before the pass start date.",
        )
    return valid_until


def _effective_status(bus_pass: BusPass) -> str:
    stored = (bus_pass.status or "Pending").strip().title()
    if stored == "Active" and bus_pass.valid_until and bus_pass.valid_until < date.today():
        return "Expired"
    if stored == "Active" and bus_pass.valid_from and bus_pass.valid_from > date.today():
        return "Pending"
    return stored


def _assignment(student: Student, db: Session) -> tuple[Route | None, Bus | None]:
    route = student.route or (
        db.query(Route).filter(Route.bus_id == student.bus_id).first()
        if student.bus_id else None
    )
    bus = db.get(Bus, route.bus_id) if route and route.bus_id else (
        student.bus if route is None else None
    )
    return route, bus


def _serialize(student: Student, bus_pass: BusPass | None, db: Session) -> dict:
    route, bus = _assignment(student, db)
    today = date.today()
    effective_status = _effective_status(bus_pass) if bus_pass else None
    days_until_expiry = (
        (bus_pass.valid_until - today).days
        if bus_pass and bus_pass.valid_until else None
    )
    return {
        "student": {
            "id": student.id,
            "name": student.user.full_name if student.user else "Unknown student",
            "student_code": student.student_code,
        },
        "transport": {
            "bus_number": bus.bus_number if bus else None,
            "route_name": route.route_name if route else None,
            "boarding_stop": student.stop.stop_name if student.stop else None,
            "assigned": bool(route and bus),
        },
        "bus_pass": (
            {
                "id": bus_pass.id,
                "pass_number": bus_pass.pass_number,
                "status": bus_pass.status,
                "effective_status": effective_status,
                "valid_from": bus_pass.valid_from,
                "valid_until": bus_pass.valid_until,
                "validity_period": bus_pass.validity_period,
                "academic_year": bus_pass.academic_year,
                "issued_at": bus_pass.issued_at,
                "days_until_expiry": days_until_expiry,
                "expiring_soon": (
                    effective_status == "Active"
                    and days_until_expiry is not None
                    and 0 <= days_until_expiry <= 30
                ),
            }
            if bus_pass else None
        ),
    }


def _issued_pass_number(student: Student, db: Session) -> str:
    """Generate a readable pass number without exposing personal data."""

    for _ in range(10):
        candidate = f"BP-{student.student_code}-{secrets.token_hex(3).upper()}"
        if db.query(BusPass.id).filter(BusPass.pass_number == candidate).first() is None:
            return candidate
    raise HTTPException(status_code=500, detail="Could not generate a unique bus pass number.")


@router.get("")
def list_bus_passes(
    db: Session = Depends(get_db),
    _current_user: User = Depends(require_management),
):
    """List every student with their pass and authoritative assignment state."""

    students = db.query(Student).order_by(Student.student_code).all()
    records = [_serialize(student, student.bus_pass, db) for student in students]
    return {
        "records": records,
        "total": len(records),
        "issued": sum(record["bus_pass"] is not None for record in records),
        "expiring_soon": sum(
            bool(record["bus_pass"] and record["bus_pass"]["expiring_soon"])
            for record in records
        ),
    }


@router.post("", status_code=status.HTTP_201_CREATED)
def issue_bus_pass(
    payload: BusPassIssue,
    db: Session = Depends(get_db),
    _current_user: User = Depends(require_management),
):
    """Issue one yearly/two-semester pass for a centrally assigned student."""

    student = db.get(Student, payload.student_id)
    if student is None:
        raise HTTPException(status_code=404, detail="Student not found.")
    if student.bus_pass is not None:
        raise HTTPException(status_code=409, detail="This student already has a bus pass. Renew or update it instead.")
    route, bus = _assignment(student, db)
    if route is None or bus is None:
        raise HTTPException(
            status_code=400,
            detail="Assign the student to a route with a bus before issuing a pass.",
        )

    bus_pass = BusPass(
        student_id=student.id,
        pass_number=_issued_pass_number(student, db),
        status=payload.status,
        valid_from=payload.valid_from,
        valid_until=_resolve_valid_until(
            payload.valid_from,
            payload.validity_period,
            payload.valid_until,
        ),
        validity_period=payload.validity_period,
        academic_year=payload.academic_year.strip() if payload.academic_year else None,
        issued_at=datetime.now(timezone.utc),
    )
    db.add(bus_pass)
    db.commit()
    db.refresh(bus_pass)
    return _serialize(student, bus_pass, db)


@router.put("/{pass_id:int}")
def update_bus_pass(
    pass_id: int,
    payload: BusPassUpdate,
    db: Session = Depends(get_db),
    _current_user: User = Depends(require_management),
):
    """Renew a pass from a new start date or change its active/suspended state."""

    bus_pass = db.get(BusPass, pass_id)
    if bus_pass is None:
        raise HTTPException(status_code=404, detail="Bus pass not found.")

    next_valid_from = payload.valid_from or bus_pass.valid_from
    next_validity_period = payload.validity_period or bus_pass.validity_period
    dates_changed = payload.valid_from is not None or payload.valid_until is not None
    if payload.validity_period is not None and payload.validity_period != bus_pass.validity_period:
        dates_changed = True

    if dates_changed:
        if next_valid_from is None:
            raise HTTPException(status_code=422, detail="A pass start date is required.")
        bus_pass.valid_from = next_valid_from
        bus_pass.valid_until = _resolve_valid_until(
            next_valid_from,
            next_validity_period,
            payload.valid_until,
        )
        bus_pass.issued_at = datetime.now(timezone.utc)
    bus_pass.validity_period = next_validity_period
    bus_pass.academic_year = payload.academic_year.strip() if payload.academic_year else None
    bus_pass.status = payload.status

    db.commit()
    db.refresh(bus_pass)
    return _serialize(bus_pass.student, bus_pass, db)


@router.delete("/{pass_id:int}", status_code=status.HTTP_204_NO_CONTENT)
def delete_bus_pass(
    pass_id: int,
    db: Session = Depends(get_db),
    _current_user: User = Depends(require_management),
):
    """Remove an issued pass without changing the student's bus assignment."""

    bus_pass = db.get(BusPass, pass_id)
    if bus_pass is None:
        raise HTTPException(status_code=404, detail="Bus pass not found.")
    db.delete(bus_pass)
    db.commit()
